### PR TITLE
chore: implement JSON Schema to NodeType adapter

### PR DESCRIPTION
**Pull request by tatomyr**
## What/Why/How?

Implemented JSON Schema to NodeType adapter.

This streamlines merging external config schemas while allowing deriving  TypeScript types from JSON Schema schemas.

The adapter itself is [here](https://github.com/Redocly/redocly-cli/pull/1401/files#diff-c4d554b08b03755267606a207150f2f1d9d293aed5875e65e1d6137f1a44986eR1). It uses AJV to determine the correct `oneOf` schema if there's no `discriminator` defined.

The adapter is limited to JSON Schemas that are not booleans or arrays themselves, nor types are arrays.
Also, the schemas should not contain `allOf` or `anyOf` (we're not using them anyway).
~The most crucial limitation is that an array's `items` could not contain `oneOf`. I even had to change the [adapters schema](https://github.com/Redocly/redocly-cli/pull/1401/files#diff-60824f92690be457bba7302950d7eba603e6580fd3850df3e49f2e6f778146b1R275) schema of the `developerOnboarding` section. That is because `NodeType` cannot alter items (like it can do with  `properties` or `additionalProperties`).~ -- Upd: fixed in https://github.com/Redocly/redocly-cli/pull/1401/commits/2fe3a1ce8fd46a4f1e0d7d28cd74d186504df3de

---

I also fixed a wrong snapshot test, added missing configurable rules subject types, and updated the config schema to the latest version.

### Performance

Investigated performance losses. It doesn't do that bad on linting. For the current Rebilly API description it doesn't show any significant difference comparing to other versions:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `node node_modules/cli-140/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.471 ± 0.035 | 1.431 | 1.554 | 1.00 ± 0.03 |
| `node node_modules/cli-150/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.470 ± 0.022 | 1.430 | 1.501 | 1.00 ± 0.02 |
| `node node_modules/cli-160/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.468 ± 0.017 | 1.454 | 1.511 | 1.00 |
| `node node_modules/cli-170/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.514 ± 0.023 | 1.481 | 1.547 | 1.03 ± 0.02 |
| `node node_modules/cli-180/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.492 ± 0.020 | 1.452 | 1.514 | 1.02 ± 0.02 |
| `node node_modules/cli-next/bin/cli.js lint api-definitions/openapi/openapi.yaml` | 1.496 ± 0.025 | 1.464 | 1.537 | 1.02 ± 0.02 |

On the other hand, for smaller museum.yaml API description the losses are higher (which most likely means the config itself gets resolved longer):

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `node node_modules/cli-140/bin/cli.js lint ../resources/museum.yaml` | 469.3 ± 11.5 | 457.2 | 491.4 | 1.00 |
| `node node_modules/cli-150/bin/cli.js lint ../resources/museum.yaml` | 472.4 ± 5.9 | 464.5 | 482.1 | 1.01 ± 0.03 |
| `node node_modules/cli-160/bin/cli.js lint ../resources/museum.yaml` | 479.0 ± 13.3 | 467.3 | 504.1 | 1.02 ± 0.04 |
| `node node_modules/cli-170/bin/cli.js lint ../resources/museum.yaml` | 473.1 ± 5.0 | 467.2 | 482.0 | 1.01 ± 0.03 |
| `node node_modules/cli-180/bin/cli.js lint ../resources/museum.yaml` | 485.0 ± 4.0 | 477.5 | [489](https://github.com/Redocly/redocly-cli/actions/runs/7883556561/job/21510885819#step:6:490).3 | 1.03 ± 0.03 |
| `node node_modules/cli-next/bin/cli.js lint ../resources/museum.yaml` | [490](https://github.com/Redocly/redocly-cli/actions/runs/7883556561/job/21510885819#step:6:491).0 ± 3.2 | 485.1 | 495.6 | 1.04 ± 0.03 |

## Check yourself

- [x] Code is linted
- [x] Tested with redoc/reference-docs/workflows (internal)
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/lint.ts [hunk 1]
```diff
@@ -140,6 +141,7 @@ export async function lintConfig(opts: {
       ruleId: 'configuration no-unresolved-refs',
       visitor: NoUnresolvedRefs({ severity: 'error' }),
     },
+    // TODO: ? use another rule similar to no-invalid-media-type-examples to validate { type: array, items: { oneOf: [...] } } like in `devOnboardingAdapterConfigSchema`?
```
Comment (RomanHotsiy) [inline packages/core/src/lint.ts:144]: can it be removed?

File: packages/core/src/types/redocly-yaml.ts [hunk 2]
```diff
@@ -261,13 +314,14 @@ const ConfigHTTP: NodeType = {
...
-const ConfigRootTheme: NodeType = {
+const createConfigRootTheme = (nodeTypes: Record<string, NodeType>): NodeType => ({
+  ...nodeTypes['rootRedoclyConfigSchema.theme'],
   properties: {
-    ...themeConfigSchema.properties,
-    openapi: 'ConfigReferenceDocs',
-    mockServer: 'ConfigMockServer',
+    ...nodeTypes['rootRedoclyConfigSchema.theme']?.properties,
+    openapi: 'ConfigReferenceDocs', // Override theme.openapi with internal format
+    // mockServer: 'ConfigMockServer', // TODO: do we still need this? It looks like we declare it in the redoclyConfigSchema
```
[the remark this answers is not in the document]
Comment (tatomyr) [reply packages/core/src/types/redocly-yaml.ts:322]: It seems like some types are outdated (like this one) or incorrect. Will cleanup what I can.

File: packages/core/src/types/json-schema-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    (isPlainObject(schema.items.properties) ||
+      isPlainObject(schema.items.additionalProperties) ||
+      schema.items.oneOf) // exclude scalar array types
+  ) {
+    items = transformJSONSchemaToNodeType(propertyName + '_items', schema.items, ctx);
+  }
+
+  let required = schema.required as NodeType['required'];
+  // Translate required in oneOfs into a ResolveTypeFn.
+  if (schema.oneOf && schema.oneOf.every((option) => !!(option as Oas3Schema).required)) {
```
Comment (IgorKarpiuk) [inline packages/core/src/types/json-schema-adapter.ts:188]: `option.required` can be an empty array?

Comment (tatomyr) [reply packages/core/src/types/json-schema-adapter.ts:188]: technically yes

**Review by IgorKarpiuk — APPROVED** [review]
LGTM

**Pull request merged**